internal/middleware/etag_middleware.go

```diff
@@ -0,0 +1,49 @@
+package middleware
+
+import (
+	"errors"
+	"fmt"
+	"hash/crc64"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/RHEnVision/provisioning-backend/internal/ctxval"
+)
+
+var InvalidETagErr = errors.New("empty etag provided")
+
+func ETagMiddleware(expires time.Duration, name, etag string) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		fn := func(w http.ResponseWriter, r *http.Request) {
+			if etag == "" {
+				panic(InvalidETagErr)
+			}
+			etagValue := fmt.Sprintf("\"pb-%s-%s\"", name, etag)
+			controlValue := fmt.Sprintf("max-age=%d", int(expires.Seconds()))
+			w.Header().Set("ETag", etagValue)
+			w.Header().Set("Cache-Control", controlValue)
+			logger := ctxval.Logger(r.Context()).With().Str("etag", etag).Str("etag_name", name).Logger()
+			logger.Trace().Msgf("Returned etag with Cache-Control '%s'", controlValue)
+
+			if match := r.Header.Get("If-None-Match"); match != "" {
+				if strings.Contains(match, etag) {
```

Comment:
Shouldn't this be exact match? I know it's a minor thing, but I'm interested, why? Some whitespace issues?

Reply:
_The If-None-Match HTTP request header makes the request conditional. For [GET](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET) and [HEAD](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/HEAD) methods, the server will return the requested resource, with a [200](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/200) status, only if it doesn't have an [ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) matching the given ones. For other methods, the request will be processed only if the eventually existing resource's [ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) doesn't match any of the values listed._

From: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match

Reply:
ILT, thanks! (this is exactly what I meant by I don't fully understand how it should work, so it's hard to review) :)